### PR TITLE
Sort for tracks in my places

### PR DESCRIPTION
**Pull request by DmitrySvetlichny**

### CODE REVIEW

File: Sources/Controllers/MyPlaces/TracksSortModeHelper.swift [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        case .shorterDurationFirst: return .icCustomSortDurationShortToLong
+        }
+    }
+    
+    static func getByTitle(_ title: String) -> TracksSortMode {
+        return TracksSortMode.allCases.first(where: { $0.title == title }) ?? .lastModified
+    }
+}
+
+@objc class TracksSortModeHelper: NSObject {
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksSortModeHelper.swift:63]: final

File: Sources/Controllers/MyPlaces/TracksSortModeHelper.swift [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        case .oldestDateFirst: return .icCustomSortDateOldest
+        case .longestDistanceFirst: return .icCustomSortLongToShort
+        case .shortestDistanceFirst: return .icCustomSortShortToLong
+        case .longestDurationFirst: return .icCustomSortDurationLongToShort
+        case .shorterDurationFirst: return .icCustomSortDurationShortToLong
+        }
+    }
+    
+    static func getByTitle(_ title: String) -> TracksSortMode {
+        return TracksSortMode.allCases.first(where: { $0.title == title }) ?? .lastModified
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksSortModeHelper.swift:59]: without return

File: Sources/Controllers/MyPlaces/TracksSortModeHelper.swift [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    }
+}
+
+@objc class TracksSortModeHelper: NSObject {
+    @objc static var defaultSortModeTitle: String {
+        return TracksSortMode.lastModified.title
+    }
+    
+    @objc static func title(for mode: TracksSortMode) -> String {
+        return mode.title
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksSortModeHelper.swift:69]: up

File: Sources/Controllers/MyPlaces/TracksSortModeHelper.swift [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+            return tracks.sorted { $0.totalDistance < $1.totalDistance }
+        case .longestDurationFirst:
+            return tracks.sorted { $0.timeSpan > $1.timeSpan }
+        case .shorterDurationFirst:
+            return tracks.sorted { $0.timeSpan < $1.timeSpan }
+        }
+    }
+    
+    static func getTrackDescription(track: GpxDataItem, sortMode: TracksSortMode, includeFolderInfo: Bool = false) -> NSAttributedString {
+        let dateFormatter = DateFormatter()
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksSortModeHelper.swift:98]: use cashed DateFormatter

File: Sources/Controllers/MyPlaces/TracksSortModeHelper.swift [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        
+        return fullString
+    }
+    
+    static func distanceToGPX(gpx: GpxDataItem) -> CGFloat {
+        guard let currentLocation = OsmAndApp.swiftInstance().locationServices?.lastKnownLocation else {
+            return CGFloat.greatestFiniteMagnitude
+        }
+        
+        if let analysis = gpx.getAnalysis(), let start = analysis.getLatLonStart(), CLLocationCoordinate2DIsValid(CLLocationCoordinate2DMake(start.latitude, start.longitude)) {
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksSortModeHelper.swift:177]: guard

File: Sources/Controllers/MyPlaces/TracksViewController.swift [hunk 6]
```diff
@@ -554,11 +651,26 @@ final class TracksViewController: OACompoundViewController, UITableViewDelegate,
         toolbarItems = [selectDeselectButton]
     }
     
-    // #warning("implement sorting in next task")  // See: https://github.com/osmandapp/OsmAnd-Issues/issues/2348
-    private func sortWithOptions(_ list: [String], options: SortingOptions) -> [String] {
+    private func sortFolderWithMode(_ list: [String], mode: TracksSortMode) -> [String] {
         list.sorted { $0 < $1 }
     }
     
+    func updateDistanceAndDirection(_ forceUpdate: Bool) {
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksViewController.swift:658]: private?

File: Sources/Controllers/MyPlaces/TracksViewController.swift [hunk 7]
```diff
@@ -554,11 +651,26 @@ final class TracksViewController: OACompoundViewController, UITableViewDelegate,
...
+        lock.lock()
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksViewController.swift:659]: ?

File: Sources/Controllers/MyPlaces/TracksSortModeHelper.swift [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    case nameAZ
+    case nameZA
+    case newestDateFirst
+    case oldestDateFirst
+    case longestDistanceFirst
+    case shortestDistanceFirst
+    case longestDurationFirst
+    case shorterDurationFirst
+    
+    static var allCases: [TracksSortMode] {
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksSortModeHelper.swift:24]: don't use it. https://developer.apple.com/documentation/swift/caseiterable

File: Sources/Controllers/MyPlaces/TracksViewController.swift [hunk 9]
```diff
@@ -490,6 +510,83 @@ final class TracksViewController: OACompoundViewController, UITableViewDelegate,
         filterButton.configuration = currentConfig
     }
     
+    private func createSortMenu() -> UIMenu {
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksViewController.swift:513]: Move all the logic (functions) related to sorting into a separate extension. Can keep it in the same class

File: Sources/Controllers/MyPlaces/TracksSortModeHelper.swift [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+
+@objc final class TracksSortModeHelper: NSObject {
+    static let dateFormatter: DateFormatter = {
+        let formatter = DateFormatter()
+        formatter.dateFormat = "dd.MM.yyyy"
+        return formatter
+    }()
+    
+    @objc static var defaultSortModeTitle: String {
+        return TracksSortMode.lastModified.title
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksSortModeHelper.swift:67]: return

File: Sources/Controllers/MyPlaces/TracksSortModeHelper.swift [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+    
+    static func getByTitle(_ title: String) -> TracksSortMode {
+        TracksSortMode.allCases.first(where: { $0.title == title }) ?? .lastModified
+    }
+}
+
+@objc final class TracksSortModeHelper: NSObject {
+    static let dateFormatter: DateFormatter = {
+        let formatter = DateFormatter()
+        formatter.dateFormat = "dd.MM.yyyy"
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksSortModeHelper.swift:62]: native format?

File: Sources/Controllers/MyPlaces/TracksViewController.swift [hunk 12]
```diff
@@ -490,6 +509,41 @@ final class TracksViewController: OACompoundViewController, UITableViewDelegate,
         filterButton.configuration = currentConfig
     }
     
+    private func updateSortButtonAndMenu() {
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksViewController.swift:512]: move to extension

File: Sources/Controllers/MyPlaces/TracksSortModeHelper.swift [hunk 13]
```diff
@@ -96,6 +131,19 @@ import OsmAndShared
         }
     }
     
+    static func descriptionForFolder(folder: TrackFolder, currentFolderPath: String) -> String {
+        let folderName = folder.getDirName()
+        let tracksCount = folder.totalTracksCount
+        let basicDescription = String(format: localizedString("folder_tracks_count"), tracksCount)
+        
+        if let lastModifiedDate = OAUtilities.getFileLastModificationDate(currentFolderPath.appendingPathComponent(folderName)) {
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksSortModeHelper.swift:139]: folder.lastModified() ?

File: Sources/Controllers/MyPlaces/TracksViewController.swift [hunk 14]
```diff
@@ -1363,6 +1361,22 @@ final class TracksViewController: OACompoundViewController, UITableViewDelegate,
...
     }
+    
+    private func renameSortModeKey(from oldName: String, to newName: String) {
+        var sortModes = settings.getTracksSortModes()
+        if let sortMode = sortModes?[oldName] {
+            sortModes?[newName] = sortMode
+            sortModes?.removeValue(forKey: oldName)
+        }
+        
+        settings.saveTracksSortModes(sortModes)
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksViewController.swift:1372]: Save it if it's really needed. Move it into the "if" block

File: Sources/Controllers/MyPlaces/TracksViewController.swift [hunk 15]
```diff
@@ -1363,6 +1361,22 @@ final class TracksViewController: OACompoundViewController, UITableViewDelegate,
...
+    }
+    
+    private func removeSortMode(forFolderName folderName: String) {
+        var sortModes = settings.getTracksSortModes()
+        sortModes?.removeValue(forKey: folderName)
+        settings.saveTracksSortModes(sortModes)
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksViewController.swift:1378]: up

**Review by tigrim — APPROVED** [review]

File: Sources/Controllers/MyPlaces/TracksViewController.swift [hunk 16]
```diff
@@ -1314,8 +1364,35 @@ final class TracksViewController: OACompoundViewController, UITableViewDelegate,
             }
         }
     
+        renameSortModeKey(from: folderPathForOpenedContextMenu, to: destinationShortFolderPath)
         updateAllFoldersVCData(forceLoad: true)
     }
+    
+    private func renameSortModeKey(from oldBasePath: String, to newBasePath: String) {
```
Comment (alex-osm) [inline Sources/Controllers/MyPlaces/TracksViewController.swift:1371]: what purpose of this method?

**Pull request merged**